src/grammar.rs

```diff
@@ -48,19 +52,111 @@ impl Grammar {
         }
     }
 
-    /// Get iterator of the `Grammar`'s `Productions`s
+    /// Get iterator of the `Grammar`'s `Production`s
     pub fn productions_iter(&self) -> Iter {
         Iter {
             iterator: self.productions.iter(),
         }
     }
 
-    /// Get mutable iterator of the `Grammar`'s `Productions`s
+    /// Get mutable iterator of the `Grammar`'s `Production`s
     pub fn productions_iter_mut(&mut self) -> IterMut {
         IterMut {
             iterator: self.productions.iter_mut(),
         }
     }
+
+    fn eval_terminal(&self, term: &Term) -> Result<String, Error> {
+        match *term {
+            Term::Nonterminal(ref nt) => self.traverse(&nt),
+            Term::Terminal(ref t) => Ok(t.clone()),
+        }
+    }
+
+    fn traverse(&self, ident: &String) -> Result<String, Error> {
+        let stack_red_zone: usize = 32 * 1024; // 32KB
+        if stacker::remaining_stack() <= stack_red_zone {
+            return Err(Error::GenerateError(
+                String::from("Infinite loop detected!"),
+            ));
+        }
+
+        let nonterm = Term::Nonterminal(ident.clone());
+        let mut result = String::new();
+        let production;
+        let find_lhs = self.productions_iter().find(|&x| x.lhs == nonterm);
+
+        match find_lhs {
+            Some(p) => production = p,
+            None => return Ok(nonterm.to_string()),
+        }
+
+        let expression;
+        let expressions = production.rhs_iter().collect::<Vec<&Expression>>();
+
+        match thread_rng().choose(&expressions) {
+            Some(e) => expression = e.clone(),
+            None => {
+                return Err(Error::GenerateError(
+                    String::from("Couldn't select random Expression!"),
+                ));
+            }
+        }
+
+        for term in expression.terms_iter() {
+            match self.eval_terminal(&term) {
+                Ok(s) => result = result + &s,
+                Err(e) => return Err(e),
+            }
+        }
+
+        Ok(result)
+    }
+
+    /// Generate a random sentence from self.
+    /// Begins from lhs of first production.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// extern crate bnf;
+    /// use bnf::Grammar;
+    ///
+    /// fn main() {
+    ///     let input =
+    ///         "<dna> ::= <base> | <base> <dna>
+    ///         <base> ::= \"A\" | \"C\" | \"G\" | \"T\"";
+    ///     let grammar = Grammar::from_str(input).unwrap();
+    ///     let sentence = grammar.generate();
+    ///     match sentence {
+    ///         Ok(s) => println!("random sentence: {}", s),
+    ///         Err(e) => println!("something went wrong: {}!", e)
+    ///     }
+    /// }
+    /// ```
+    pub fn generate(&self) -> Result<String, Error> {
+        let start_rule: String;
+        let lhs = self.productions_iter().nth(0);
```

Comment:
This variable is the first production, not the lhs as the name suggests. And some of the name confusion continues in the match below, referring to the production as a term.